Userland/Utilities/service.cpp

```diff
@@ -0,0 +1,92 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Format.h>
+#include <AK/StringView.h>
+#include <LibCore/ArgsParser.h>
+#include <LibIPC/ServerConnection.h>
+#include <SystemServer/ClientConnection.h>
+#include <SystemServer/ServiceManagementClientEndpoint.h>
+#include <SystemServer/ServiceManagementServerEndpoint.h>
+
+class Client final
+    : public IPC::ServerConnection<ServiceManagementClientEndpoint, ServiceManagementServerEndpoint>
+    , public ServiceManagementClientEndpoint {
+    C_OBJECT(Client);
+
+public:
+    virtual void handshake() override;
+
+private:
+    Client();
+
+    virtual void handle(const Messages::ServiceManagementClient::Dummy&) override;
+};
+
+Client::Client()
+    : IPC::ServerConnection<ServiceManagementClientEndpoint, ServiceManagementServerEndpoint>(*this, "/tmp/portal/system")
+{
+    handshake();
+}
+
+void Client::handshake()
+{
+    send_sync<Messages::ServiceManagementServer::Greet>();
+}
+
+void Client::handle(const Messages::ServiceManagementClient::Dummy&)
+{
+}
+
+int main(int argc, char** argv)
+{
+    const char* service_name;
+    const char* command;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(service_name, "Service to manage", "service", Core::ArgsParser::Required::Yes);
+    parser.add_positional_argument(command, "Command", "command", Core::ArgsParser::Required::Yes);
+    if (!parser.parse(argc, argv))
+        return 1;
+
+    auto command_string = StringView(command);
+
+    Core::EventLoop loop;
+    auto connection = Client::construct();
+
+    dbgln("$ service {} {}", service_name, command);
+
+    if (command_string == "status"sv) {
+        auto status = connection->send_sync<Messages::ServiceManagementServer::ServiceStatus>(service_name);
+        outln("{}", status->status());
+        loop.quit(0);
```

Comment:
Does it ever actually run the loop? It appears to always `return 1` below.

Reply:
This is removed in some next commit.